packages/@ember/-internals/routing/lib/system/router.ts

```diff
@@ -1438,7 +1438,11 @@ function calculatePostTransitionState(
 
     // If the routeInfo is not resolved, we serialize the context into params
     if (!routeInfo.isResolved) {
-      params[routeInfo.name] = routeInfo.serialize(routeInfo.context);
+      params[routeInfo.name] = routeInfo.serialize(routeInfo.context as
```

Comment:
@rwjblue seems `serialize` accepting `null` is okay, should we change the types in router_js or just cast here?

Reply:
What does `null` mean in that context?

Either way, if it is a supported value in the router_js code, we ought to type it that way...

Reply:
It traces back to `Route.model` can return `null` https://github.com/tildeio/router.js/blob/v7.0.0/lib/router/route-info.ts#L26, the model can be assigned to `routeInfo.context` https://github.com/tildeio/router.js/blob/v7.0.0/lib/router/route-info.ts#L472

Reply:
Sure, but `TransitionState.prototype.params` is just this:

```ts
interface IParams {
  [key: string]: unknown;
}
```

https://github.com/tildeio/router.js/blob/cfe549c68aca9f14e834658e12b2633c435198db/lib/router/transition-state.ts#L7-L10

Why wouldn't `unknown allow `null` there? I assume we must be narrowing incorrectly here in Ember (not in `router_js`).

Reply:
`routeInfo.serialize` return type is fine here, the problem is `routeInfo.context` passed to `_context` cannot be `null`

```ts
  // https://github.com/tildeio/router.js/blob/v7.0.0/lib/router/route-info.ts#L221
  serialize(_context?: Dict<unknown> /*<- this is problematic*/): Dict<unknown> | undefined /*<- this is fine*/ {
    return this.params || {};
  }
```

The actual code in the base class `InternalRouteInfo` and inherit  `UnresolvedRouteInfoByObject` can deal with falsy `_context`
https://github.com/tildeio/router.js/blob/v7.0.0/lib/router/route-info.ts#L515

Reply:
I guess I'm a bit surprised that this has anything to do with `model` (https://github.com/tildeio/router.js/blob/cfe549c68aca9f14e834658e12b2633c435198db/lib/router/route-info.ts#L23-L26), since the `context` argument here is effectively the params (not the model result).

Regardless, you are totally correct that `route-info` itself is not internally consistent.

https://github.com/tildeio/router.js/blob/cfe549c68aca9f14e834658e12b2633c435198db/lib/router/route-info.ts#L205

And 

https://github.com/tildeio/router.js/blob/cfe549c68aca9f14e834658e12b2633c435198db/lib/router/route-info.ts#L221

Should match...

Reply:
Ugh it does not seem straight forward to make them match, I ran into promise / no-promise type error.
My guess from reading the code is that the router_js routeInfo `serialize` method only accepts resolved model or identifier (param). Its context can be these or unresolved model promise, but not inside `calculatePostTransitionState`.

Likely router_js needs a larger type refactoring around different RouteInfos but I don't have a clear idea now.

Reply:
Ya, I was thinking that you want the resolved values from https://github.com/tildeio/router.js/blob/cfe549c68aca9f14e834658e12b2633c435198db/lib/router/route-info.ts#L205

So basically:

```ts
Dict<unknown> | null | undefined
```